File: skill_manager/api/schemas/scan.py
from __future__ import annotations

from pydantic import BaseModel


class ScanOptionsRequest(BaseModel):
    useLlm: bool = True
    llmApiKey: str | None = None
    llmModel: str | None = None
    llmBaseUrl: str | None = None
    llmProvider: str | None = None
    llmApiVersion: str | None = None
    llmMaxTokens: int = 8192
    llmConsensusRuns: int = 1
    awsRegion: str | None = None
    awsProfile: str | None = None
    awsSessionToken: str | None = None


class ScanFindingResponse(BaseModel):
    id: str
    ruleId: str
    category: str
    severity: str
    title: str
    description: str
    filePath: str | None = None
    lineNumber: int | None = None
    snippet: str | None = None
    remediation: str | None = None
    analyzer: str | None = None
    metadata: dict = {}


class ScanResultResponse(BaseModel):
    skillName: str
    isSafe: bool
    maxSeverity: str
    findingsCount: int
    findings: list[ScanFindingResponse]
    analyzersUsed: list[str]
    durationSeconds: float


class ScanAvailabilityResponse(BaseModel):
    available: bool


class DetectedProviderResponse(BaseModel):
    provider: str
    apiKeySource: str
    model: str | None = None
    baseUrl: str | None = None
    isAvailable: bool


class LLMDetectionResponse(BaseModel):
    providers: list[DetectedProviderResponse]
    defaultModel: str | None = None
    defaultProvider: str | None = None
    hasAnyAvailable: bool


class ScanConfigItem(BaseModel):
    id: int
    name: str
    baseUrl: str
    apiKeyMasked: str
    model: str
    provider: str
    apiVersion: str
    awsRegion: str
    awsProfile: str
    maxTokens: int
    consensusRuns: int
    isActive: bool
    lastValidatedAt: str | None = None
    lastValidationError: str = ""


class ScanConfigSecretResponse(BaseModel):
    apiKey: str


class ScanConfigListResponse(BaseModel):
    configs: list[ScanConfigItem]
    activeId: int | None


class ScanConfigSaveRequest(BaseModel):
    name: str
    baseUrl: str
    apiKey: str
    model: str
    provider: str = ""
    apiVersion: str = ""
    maxTokens: int = 8192
    consensusRuns: int = 1
    awsRegion: str = ""
    awsProfile: str = ""
    awsSessionToken: str = ""


class ScanConfigValidateRequest(ScanConfigSaveRequest):
    existingConfigId: int | None = None


class ScanConfigValidationResponse(BaseModel):
    ok: bool
    message: str
    provider: str | None = None
    model: str | None = None
    durationMs: int | None = None
    errorCode: str | None = None

File: frontend/src/api/scan.ts
import { postJson, putJson, fetchJson, deleteJson } from "./http";

export interface ScanFinding {
  id: string;
  ruleId: string;
  category: string;
  severity: string;
  title: string;
  description: string;
  filePath: string | null;
  lineNumber: number | null;
  snippet: string | null;
  remediation: string | null;
  analyzer: string | null;
}

export interface ScanResult {
  skillName: string;
  isSafe: boolean;
  maxSeverity: string;
  findingsCount: number;
  findings: ScanFinding[];
  analyzersUsed: string[];
  durationSeconds: number;
}

export interface ScanAvailability {
  available: boolean;
}

export interface DetectedProvider {
  provider: string;
  apiKeySource: string;
  model: string | null;
  baseUrl: string | null;
  isAvailable: boolean;
}

export interface LLMDetection {
  providers: DetectedProvider[];
  defaultModel: string | null;
  defaultProvider: string | null;
  hasAnyAvailable: boolean;
}

export async function checkScanAvailability(): Promise<ScanAvailability> {
  return fetchJson<ScanAvailability>("/scan/availability");
}

export async function detectLLM(): Promise<LLMDetection> {
  return fetchJson<LLMDetection>("/scan/llm/detection");
}

export async function scanSkill(
  skillRef: string,
  options?: {
    useBehavioral?: boolean;
    useLlm?: boolean;
    llmBaseUrl?: string;
    llmApiKey?: string;
    llmModel?: string;
    llmProvider?: string;
    llmApiVersion?: string;
    llmMaxTokens?: number;
    llmConsensusRuns?: number;
    awsRegion?: string;
    awsProfile?: string;
    awsSessionToken?: string;
  },
): Promise<ScanResult> {
  return postJson<ScanResult>(
    `/scan/skills/${encodeURIComponent(skillRef)}`,
    options ?? {},
  );
}

export interface ScanConfigItem {
  id: number;
  name: string;
  baseUrl: string;
  apiKeyMasked: string;
  model: string;
  provider: string;
  apiVersion: string;
  awsRegion: string;
  awsProfile: string;
  maxTokens: number;
  consensusRuns: number;
  isActive: boolean;
  lastValidatedAt: string | null;
  lastValidationError: string;
}

export interface ScanConfigListResponse {
  configs: ScanConfigItem[];
  activeId: number | null;
}

export interface ScanConfigSecretResponse {
  apiKey: string;
}

export interface ScanConfigSaveRequest {
  name: string;
  baseUrl: string;
  apiKey: string;
  model: string;
  provider?: string;
  apiVersion?: string;
  maxTokens?: number;
  consensusRuns?: number;
  awsRegion?: string;
  awsProfile?: string;
  awsSessionToken?: string;
}

export interface ScanConfigValidateRequest extends ScanConfigSaveRequest {
  existingConfigId?: number;
}

export interface ScanConfigValidationResponse {
  ok: boolean;
  message: string;
  provider: string | null;
  model: string | null;
  durationMs: number | null;
  errorCode: string | null;
}

export async function getScanConfigs(): Promise<ScanConfigListResponse> {
  return fetchJson<ScanConfigListResponse>("/scan/configs");
}

export async function revealScanConfigApiKey(id: number): Promise<ScanConfigSecretResponse> {
  return fetchJson<ScanConfigSecretResponse>(`/scan/configs/${id}/secret`);
}

export async function createScanConfig(config: ScanConfigSaveRequest): Promise<ScanConfigItem> {
  return postJson<ScanConfigItem>("/scan/configs", config);
}

export async function updateScanConfig(id: number, config: ScanConfigSaveRequest): Promise<ScanConfigItem> {
  return putJson<ScanConfigItem>(`/scan/configs/${id}`, config);
}

export async function validateScanConfig(config: ScanConfigValidateRequest): Promise<ScanConfigValidationResponse> {
  return postJson<ScanConfigValidationResponse>("/scan/configs/validate", config);
}

export async function deleteScanConfig(id: number): Promise<void> {
  await deleteJson(`/scan/configs/${id}`);
}

export async function setActiveScanConfig(id: number): Promise<void> {
  await putJson(`/scan/configs/${id}/active`, {});
}
